### DISCUSSION TITLE
Create activity properties in runtime.

### DISCUSSION THREAD

**Discussion by tomy2105** [elsa-workflows/elsa-core#875] [Q&A]
My use case is that at initialization I need to discover number of properties and property types for an activity.

Since properties, at least at first glance, seem to be tied to activity class members this type of dynamic adding seems impossible or would require dynamic creation of the whole class or something equally complex.

Is something like this possible? Did I miss some relatively easy way of doing this?

One idea that came to my mind, would it be possible to have several activity properties (as displayed in designer) map to one activity member variable (map, array, complex object with submemeber)? One property going to member.first or member["first"] and other going to member.second or member["second"]?

**Comment by sfmskywalker** [comment]
You should be able to do it today with relative ease. The key ingredient for this is the `IActivityTypeProvider` abstraction. When you implement this interface (and register it with DI), Elsa will use it to discover activities.

As part of this implementation, you construct a thing called an `ActivityType`, which contains a description of the activity. This description includes a list of activity properties that you can set. 

Elsa contains an implementation of `IActivityTypeProvider` that uses the activity classes to dynamically construct these activity types. As you can see, activity classes are merely a "source" of material to generate activity types from. You're free to add more activity type providers. Two other such examples are the `NotificationActivityTypeProvider` (from the Telnyx activity library that yields activity types based on Telnyx webhook payload classes) and `WebhookActivityTypeProvider`, which dynamiclly yield activity types based on webhook definitions stored in the database.

We'll have more activity providers in the future, such as one that use Open API and GraphQL schemas to dynamically yield activity types that represent various API actions.

**Reply by tomy2105** [reply]
Thanks for the verbose answer.
I haven't had time to try this so will let you know when I do.

Looking at the code I only have one question. Correct me if anything below is wrong.
When I create `ActivityType` it has `Describe` which returns `ActivityDescriptor` which in turns has `Properties` which is array of `ActivityPropertyDescriptor`. This will be properties of an Activity displayed in designer.

`ActivityPropertyDescriptor` has `Name` and `Type` and other members used to render it in designer.
My question is how does Elsa Core map `ActivityPropertyDescriptor` to an `Activity` member variable that it sets?
Full name mapping?  I need to have `Activity` member variable with identical name as `ActivityPropertyDescriptor` `Name`?
Or can I have a `ActivityPropertyDescriptor` with `Name` equal to `MySet["foo"]` and the elsa core will set this property value to the `MySet` member of my `Activity` derived class with key `foo`?

**Comment by tomy2105** [comment]
I believe I've found the solution to my problem, for my code see https://github.com/elsa-workflows/elsa-core/issues/894.
If there is anything in my solution that I should not be doing, or will not be supported in the future, please shout :)
